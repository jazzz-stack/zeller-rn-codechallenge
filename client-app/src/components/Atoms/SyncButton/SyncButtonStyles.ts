import {StyleSheet} from 'react-native';

export const styles = StyleSheet.create({
  syncButton: {
    flexDirection: 'row',
    alignItems: 'center',
    justifyContent: 'center',
    backgroundColor: '#fff',
    paddingVertical: 12,
    paddingHorizontal: 16,
    borderRadius: 8,
    borderWidth: 1,
    borderColor: '#3498DB',
    elevation: 2,
    shadowColor: '#000',
    shadowOffset: {width: 0, height: 2},
    shadowOpacity: 0.1,
    shadowRadius: 4,
  },
  disabledButton: {
    backgroundColor: '#f5f5f5',
    borderColor: '#ddd',
    elevation: 0,
    shadowOpacity: 0,
  },
  syncButtonText: {
    color: '#3498DB',
    fontSize: 16,
    fontWeight: '600',
    marginLeft: 8,
  },
  disabledText: {
    color: '#999',
  },
  spinning: {
    // Note: For actual spinning animation, you'd need to use Animated API
    // This is just a placeholder for the spinning effect
  },
});
