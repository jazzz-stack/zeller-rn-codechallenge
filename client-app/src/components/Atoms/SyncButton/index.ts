export {default} from './SyncButton';
export type {SyncButtonProps} from './SyncButton';
