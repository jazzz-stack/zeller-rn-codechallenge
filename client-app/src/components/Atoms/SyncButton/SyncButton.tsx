import React from 'react';
import {TouchableOpacity, TouchableOpacityProps} from 'react-native';
import Icon from 'react-native-vector-icons/MaterialIcons';
import Text from '../Text';
import {styles} from './SyncButtonStyles';

export interface SyncButtonProps extends Omit<TouchableOpacityProps, 'style'> {
  title?: string;
  isLoading?: boolean;
  disabled?: boolean;
  iconSize?: number;
  iconColor?: string;
  containerStyle?: any;
  textStyle?: any;
}

const SyncButton: React.FC<SyncButtonProps> = ({
  title = 'Sync',
  isLoading = false,
  disabled = false,
  iconSize = 25,
  iconColor = '#3498DB',
  containerStyle,
  textStyle,
  ...touchableProps
}) => {
  const isDisabled = disabled || isLoading;

  return (
    <TouchableOpacity
      style={[
        styles.syncButton,
        isDisabled && styles.disabledButton,
        containerStyle,
      ]}
      disabled={isDisabled}
      {...touchableProps}
    >
      <Icon
        name={isLoading ? 'refresh' : 'sync'}
        size={iconSize}
        color={isDisabled ? '#999' : iconColor}
        style={isLoading ? styles.spinning : undefined}
      />
      <Text style={[styles.syncButtonText, isDisabled && styles.disabledText, textStyle]}>
        {isLoading ? 'Syncing...' : title}
      </Text>
    </TouchableOpacity>
  );
};

export default SyncButton;
