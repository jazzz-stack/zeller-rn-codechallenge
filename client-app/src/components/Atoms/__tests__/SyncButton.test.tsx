import React from 'react';
import {render, fireEvent, screen} from '@testing-library/react-native';
import SyncButton from '../SyncButton/SyncButton';

// Mock react-native-vector-icons
jest.mock('react-native-vector-icons/MaterialIcons', () => {
  const React = require('react');
  const {Text} = require('react-native');
  
  return ({name, size, color, style, testID}: any) =>
    React.createElement(Text, {
      testID: testID || 'mock-icon',
      style: [style, {fontSize: size, color}],
    }, name);
});

describe('SyncButton', () => {
  const defaultProps = {
    onPress: jest.fn(),
  };

  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('renders correctly in different states', () => {
    // Default state
    const {rerender} = render(<SyncButton {...defaultProps} />);
    expect(screen.getByText('Sync')).toBeTruthy();
    expect(screen.getByText('sync')).toBeTruthy();

    // Loading state
    rerender(<SyncButton {...defaultProps} isLoading={true} />);
    expect(screen.getByText('Syncing...')).toBeTruthy();
    expect(screen.getByText('refresh')).toBeTruthy();

    // Custom title
    rerender(<SyncButton {...defaultProps} title="Retry" />);
    expect(screen.getByText('Retry')).toBeTruthy();
  });

  it('handles user interactions correctly', () => {
    const {rerender} = render(<SyncButton {...defaultProps} />);
    
    // Normal press
    fireEvent.press(screen.getByText('Sync').parent);
    expect(defaultProps.onPress).toHaveBeenCalledTimes(1);

    // Disabled state - should not call onPress
    rerender(<SyncButton {...defaultProps} disabled={true} />);
    fireEvent.press(screen.getByText('Sync').parent);
    expect(defaultProps.onPress).toHaveBeenCalledTimes(1); // Still 1, not called again

    // Loading state - should not call onPress
    rerender(<SyncButton {...defaultProps} isLoading={true} />);
    fireEvent.press(screen.getByText('Syncing...').parent);
    expect(defaultProps.onPress).toHaveBeenCalledTimes(1); // Still 1, not called again
  });

  it('applies custom styles and props correctly', () => {
    const customStyle = {backgroundColor: 'red'};
    const customTextStyle = {fontSize: 20};
    const testID = 'sync-button';
    
    render(
      <SyncButton 
        {...defaultProps} 
        containerStyle={customStyle}
        textStyle={customTextStyle}
        testID={testID}
        iconSize={30}
        iconColor="#FF0000"
      />
    );

    // Check container style
    const button = screen.getByTestId(testID);
    expect(button.props.style).toEqual(expect.objectContaining(customStyle));

    // Check icon props
    const icon = screen.getByText('sync');
    expect(icon.props.style).toEqual(
      expect.arrayContaining([
        expect.objectContaining({
          fontSize: 30,
          color: '#FF0000',
        }),
      ])
    );

    // Check text style
    const text = screen.getByText('Sync');
    const flattenedStyle = JSON.stringify(text.props.style);
    expect(flattenedStyle).toContain(JSON.stringify(customTextStyle));
  });

  it('handles disabled styling correctly', () => {
    const {rerender} = render(<SyncButton {...defaultProps} disabled={true} />);
    
    // Disabled state
    let icon = screen.getByText('sync');
    expect(icon.props.style).toEqual(
      expect.arrayContaining([
        expect.objectContaining({color: '#999'}),
      ])
    );

    // Loading state (also disabled)
    rerender(<SyncButton {...defaultProps} isLoading={true} />);
    icon = screen.getByText('refresh');
    expect(icon.props.style).toEqual(
      expect.arrayContaining([
        expect.objectContaining({color: '#999'}),
      ])
    );
  });

  it('handles accessibility and edge cases', () => {
    const accessibilityLabel = 'Sync data button';
    const {rerender} = render(
      <SyncButton {...defaultProps} accessibilityLabel={accessibilityLabel} />
    );
    
    // Accessibility
    expect(screen.getByLabelText(accessibilityLabel)).toBeTruthy();

    // Empty title
    rerender(<SyncButton {...defaultProps} title="" />);
    expect(screen.queryByText('Sync')).toBeNull();
    expect(screen.getByText('sync')).toBeTruthy(); // Icon still there

    // Zero icon size edge case
    rerender(<SyncButton {...defaultProps} iconSize={0} />);
    const icon = screen.getByText('sync');
    expect(icon.props.style).toEqual(
      expect.arrayContaining([
        expect.objectContaining({fontSize: 0}),
      ])
    );
  });

  it('matches snapshot', () => {
    const component = render(<SyncButton {...defaultProps} />);
    expect(component.toJSON()).toMatchSnapshot();
  });
});