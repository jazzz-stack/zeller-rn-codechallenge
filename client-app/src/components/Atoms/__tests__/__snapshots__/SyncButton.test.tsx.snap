// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SyncButton matches snapshot 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": false,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "alignItems": "center",
      "backgroundColor": "#fff",
      "borderColor": "#3498DB",
      "borderRadius": 8,
      "borderWidth": 1,
      "elevation": 2,
      "flexDirection": "row",
      "justifyContent": "center",
      "opacity": 1,
      "paddingHorizontal": 16,
      "paddingVertical": 12,
      "shadowColor": "#000",
      "shadowOffset": {
        "height": 2,
        "width": 0,
      },
      "shadowOpacity": 0.1,
      "shadowRadius": 4,
    }
  }
>
  <Text
    style={
      [
        undefined,
        {
          "color": "#3498DB",
          "fontSize": 25,
        },
      ]
    }
    testID="mock-icon"
  >
    sync
  </Text>
  <Text
    style={
      [
        {
          "fontFamily": "System",
        },
        {
          "fontSize": 16,
          "lineHeight": 24,
        },
        {
          "color": "#000000",
        },
        {
          "fontSize": 16,
        },
        {
          "fontWeight": "normal",
        },
        [
          {
            "color": "#3498DB",
            "fontSize": 16,
            "fontWeight": "600",
            "marginLeft": 8,
          },
          false,
          undefined,
        ],
      ]
    }
  >
    Sync
  </Text>
</View>
`;
